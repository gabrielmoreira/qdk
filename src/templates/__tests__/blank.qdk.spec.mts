import {
  printFsTree,
  QdkAppConfigFile,
  QdkAppConstructor,
  reset,
  resetFilesystem,
  toSnapshot,
} from '#test/helpers.mjs';
import { beforeEach, describe, expect, it, vi, vitest } from 'vitest';

vitest.mock('fs');
vitest.mock('fs/promises');

vitest.mock('#@/system/execution.mjs', () => {
  //const actual = await vi.importActual('../../system/execution.mjs');
  return {
    processCwd: vi.fn().mockReturnValue('/'),
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    execSync: vi.fn().mockImplementation((...args) => {
      // if (args[0].startsWith('npm view')) {
      //   console.log(args[0]);
      //   return actual.execSync(...args);
      // }
      return '9.9.9-mock-latest';
    }),
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    exec: vi.fn().mockImplementation(_data => {
      return '9.9.9-mock-latest';
    }),
  };
});

const importQdkConfig = async () => {
  const { default: QdkAppClass } = await vi.importActual<QdkAppConfigFile>(
    '../../templates/blank/qdk.config.mts',
  );
  return { QdkAppClass };
};

describe('blank template', () => {
  let QdkAppClass: QdkAppConstructor;

  beforeEach(async () => {
    reset();
    QdkAppClass = (await importQdkConfig()).QdkAppClass;
  });

  it('synthetize a blank template', async () => {
    // When
    await new QdkAppClass({ cwd: '/' }).synth();

    // Then
    const filesystemTree = printFsTree();
    const filesystemContent = toSnapshot();
    expect(filesystemTree).toMatchSnapshot();
    expect(filesystemContent).toMatchSnapshot();
  });

  it('synthetize a blank template and delete orphan files', async () => {
    // Given some preexistent files
    resetFilesystem({
      json: {
        '/build/readme/.qdk/meta.json': JSON.stringify({
          files: ['.qdk/meta.json', './orphan.txt'],
        }),
        '/build/readme/orphan.txt': 'some file data',
      },
    });

    // When
    await new QdkAppClass({ cwd: '/' }).synth();
    // Then
    const filesystemTree = printFsTree();
    const filesystemContent = toSnapshot();
    expect(filesystemTree).toMatchSnapshot();
    expect(filesystemContent).toMatchSnapshot();
  });
});
