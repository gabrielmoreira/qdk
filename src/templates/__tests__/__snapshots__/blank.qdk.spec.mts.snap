// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`blank template > synthetize a blank template 1`] = `
"/
└─ .qdk/
   ├─ meta.json
   └─ tree.txt"
`;

exports[`blank template > synthetize a blank template 2`] = `
{
  "/.qdk/meta.json": "{
  "project": "qdk-sample",
  "custom": {},
  "files": ["./.qdk/meta.json", "./.qdk/tree.txt"],
  "subprojects": []
}
",
  "/.qdk/tree.txt": "·
└── Project(qdk-sample)
    └── JsonFile(.qdk/meta.json)",
}
`;

exports[`blank template > synthetize a blank template and delete orphan files 1`] = `
"/
├─ .qdk/
│  ├─ meta.json
│  └─ tree.txt
└─ build/
   └─ readme/
      ├─ .qdk/
      │  └─ meta.json
      └─ orphan.txt"
`;

exports[`blank template > synthetize a blank template and delete orphan files 2`] = `
{
  "/.qdk/meta.json": "{
  "project": "qdk-sample",
  "custom": {},
  "files": ["./.qdk/meta.json", "./.qdk/tree.txt"],
  "subprojects": []
}
",
  "/.qdk/tree.txt": "·
└── Project(qdk-sample)
    └── JsonFile(.qdk/meta.json)",
  "/build/readme/.qdk/meta.json": "{"files":[".qdk/meta.json","./orphan.txt"]}",
  "/build/readme/orphan.txt": "some file data",
}
`;
